src/Disposable.js

```diff
@@ -0,0 +1,98 @@
+module.exports = function() {
+    
+    function Disposable(dispose) {
+        !!dispose && (this._dispose = dispose);
+        this.disposed = false;
+        this.length = 0;
+    }
+    
+    Disposable.prototype.dispose = function() {
+        if(!this.disposed) {
+            var result = true, f, xs, i, n;
+            (f = this._dispose) && (result = f.call(this));
+            if(!!result || result === void 0) {
+                this.disposed = true;
+                delete this._dispose;
+                while(xs = this._disposables) {
+                    i = -1;
+                    n = xs.length;
+                    delete this._disposables;
+                    while(++i < n) {
+                        xs[i].dispose();
+                    }
+                }
+                this.length = 0;
+            }
+        }
+        return result;
+    };
+    
+    Disposable.prototype.add = function () {
+        
+        var argsIdx = -1;
+        var argsLen = arguments.length;
+        var disposables = new Array(argsLen);
+        while(++argsIdx < argsLen) { disposables[argsIdx] = arguments[argsIdx]; }
+        
+        var xs = this._disposables || (this._disposables = []),
+            ys = disposables, j = xs.length, i = -1, n = ys.length,
+            xd = this.disposed, x;
+        while(++i < n) {
+            // 
+            // expression-ize all the things.
+            // 
+            // get the disposable to add
+            (x = ys[i]) &&
+            // make sure it's an object
+            (typeof x === 'object') &&
+            // make sure it isn't already disposed
+            (!x.disposed) &&
+            // make sure it has a dispose function
+            (typeof x.dispose === 'function') &&
+                // If this disposable is already disposed,
+                // immediately dispose the new disposable.
+                (xd && x.dispose() ||
+                // If this disposable isn't already disposed,
+                // or the new disposable's dispose() returns
+                // false, add it to the disposables list.
+                (xs[j++] = x));
+        }
+        this.length = j;
+        return this;
+    };
+    
+    Disposable.prototype.remove = function () {
+        
+        var argsIdx = -1;
+        var argsLen = arguments.length;
```

Comment:
Just again: Using arguments makes the function non-monomorphic. Just accept an array.